File: 64-minimum-path-sum/minimum-path-sum.java
class Solution {

    private int[][] dp;
    private int n;
    private int m;

    public int solve(int row,int col,int[][] grid)
    {
        if(row==n || col==m)
        return Integer.MAX_VALUE;

        if(row==n-1 && col==m-1)
        return grid[row][col];

        if(dp[row][col]!=-1)
        return dp[row][col];

        return dp[row][col]=grid[row][col]+Math.min(solve(row+1,col,grid),solve(row,col+1,grid));
    }

    public int minPathSum(int[][] grid) 
    {
        dp = new int[201][201];
        for (int[] row : dp)
        Arrays.fill(row,-1);

        m=grid[0].length;
        n=grid.length;

       return solve(0,0,grid);
    }
}

File: 64-minimum-path-sum/minimum-path-sum.cpp
class Solution {
public:

    int minPathSum(vector<vector<int>>& grid) 
    {
        int n=grid.size();
        int m=grid[0].size();

        vector<int> curr(m),prev(m);
        prev[0]=grid[0][0];

        for(int j=1;j<m;j++)
        prev[j]=prev[j-1]+grid[0][j];


        for(int i=1;i<n;i++)
        {
            for(int j=0;j<m;j++)
            {
                if(j==0)
                curr[j]=prev[j]+grid[i][j];
                else
                curr[j]=grid[i][j]+min(prev[j],curr[j-1]);
            }
            prev=curr;
        }
        return prev[m-1];
    }
};